src/session/session_api.c

```diff
@@ -1423,6 +1435,37 @@ __open_session(WT_CONNECTION_IMPL *conn,
 	session_ret->id = i;
 	session_ret->iface = stds;
 	session_ret->iface.connection = &conn->iface;
+	/*
+	 * Disable some methods if this is a read-only connection.  Group
+	 * them by call signature.
+	 */
+	if (F_ISSET(conn, WT_CONN_READONLY)) {
+		wt_session = &session_ret->iface;
+		wt_session->checkpoint =
+		    (int (*)(WT_SESSION *, const char *))__wt_session_notsup;
+		wt_session->compact =
+		    (int (*)(WT_SESSION *, const char *, const char *))__wt_session_notsup;
```

Comment:
I don't have enough real knowledge in this area to even be dangerous.  I certainly prefer it, stylistically.  You've convinced me.  Alex can merge once he's good.